oauthenticator/multiauthenticator.py

```diff
@@ -0,0 +1,124 @@
+"""
+Custom Authenticator to use multiple OAuth providers with JupyterHub
+
+Example of configuration:
+
+    from oauthenticator.github import GitHubOAuthenticator
+    from oauthenticator.google import GoogleOAuthenticator
+
+    c.MultiAuthenticator.authenticators = [
+        (GitHubOAuthenticator, '/github', {
+            'client_id': 'xxxx',
+            'client_secret': 'xxxx',
+            'oauth_callback_url': 'http://example.com/hub/github/oauth_callback'
+        }),
+        (GoogleOAuthenticator, '/google', {
+            'client_id': 'xxxx',
+            'client_secret': 'xxxx',
+            'oauth_callback_url': 'http://example.com/hub/google/oauth_callback'
+        }),
+        (PAMAuthenticator, "/pam", {"service_name": "PAM"}),
+    ]
+
+    c.JupyterHub.authenticator_class = 'oauthenticator.MultiAuthenticator.MultiAuthenticator'
+
+The same Authenticator class can be used several to support different providers.
+
+"""
+from jupyterhub.auth import Authenticator
+from jupyterhub.utils import url_path_join
+from traitlets import List
+
+
+class URLScopeMixin(object):
+    """Mixin class that adds the"""
+
+    scope = ""
+
+    def login_url(self, base_url):
+        return super().login_url(url_path_join(base_url, self.scope))
+
+    def logout_url(self, base_url):
+        return super().logout_url(url_path_join(base_url, self.scope))
+
+    def get_handlers(self, app):
+        handlers = super().get_handlers(app)
+        return [
+            (url_path_join(self.scope, path), handler) for path, handler in handlers
+        ]
+
+
+class MultiAuthenticator(Authenticator):
+    """Wrapper class that allows to use more than one authentication provider
+    for JupyterHub"""
+
+    authenticators = List(help="The subauthenticators to use", config=True)
+
+    def __init__(self, *arg, **kwargs):
+        super().__init__(*arg, **kwargs)
+        self._authenticators = []
+        for (
+            authenticator_klass,
+            url_scope,
+            authenticator_configuration,
+        ) in self.authenticators:
+            configuration = self.trait_values()
+            # Remove this one as it will overwrite the value if the authenticator_klass
+            # makes it configurable and the default value is used (take a look at
+            # GoogleOAuthenticator for example).
+            configuration.pop("login_service")
+
+            class WrapperAuthenticator(URLScopeMixin, authenticator_klass):
+                scope = url_scope
+
+            service_name = authenticator_configuration.pop("service_name", None)
+            configuration.update(authenticator_configuration)
+
+            authenticator = WrapperAuthenticator(**configuration)
+
```

Comment:
When providing additional config values for example (`allowed_users`) :
```
(GitHubOAuthenticator, '/github', {
  'service_name': 'github',
  'client_id': '...',
  'client_secret': '...',
  'allowed_users': {'rene.dumont'},
}),
```

Those values are not set in the created Authenticator. I am not sure for which reason. I found a way to fix this issue, see below but I am sure there is a better way linked with the way `configuration` is built.

```suggestion

            traits = authenticator.traits()
            for key, value in authenticator_configuration.items():
                trait = traits.get(key, None)
                if hasattr(authenticator, key) and trait and trait.this_class == Authenticator and trait.metadata.get('config', False):
                    setattr(authenticator, key, value)

```